### ISSUE TITLE
Extends and disabling rules (and severities)

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#26]
This is a continuation of [PR 13](https://github.com/stylelint/stylelint/pull/13) as there were a couple of things in there that I think are worth capturing for future reference.

We agreed that we shouldn't have built-in defaults, but we should instead, at some point, supply a mechanism to use or extend a recommended configuration.

If we provide an extends mechanism (like eslint does) then it follows that we might also need a means of disabling rules, right? e.g. 

```js
extends: recommended-suitcss, // example of a recommended styleguide config
rules: {
  declaration-block-trailing-semicolon: disable? 
  // the user doesn't wish to enforce either option i.e always semi or never semi
}
```

We can either: 

1. Do nothing. Which isn't bad as:
  * Power users will probably be building their own configs from scratch (maybe by copying an example recommended config from somewhere, and building on top of that).
  * New users can extend a recommend config (e.g. with a simple `extends: recommended-suitcss`) and add extra rules to it (but they can not subtract rules from it - they'll need to build their config from scratch just like the power users do, if they want to do that).
2. Do something, which has the advantages of:
  1. Easier for new users to get started.
  2. If a power user is building on top a recommended styleguide then they don't need to worry about maintaining that part of the rules configuration.

Related to this, as @MoOx pointed out in the PR, is the idea of severities e.g. it looks like eslint combines severities and disabling rules into [one number system](http://eslint.org/docs/user-guide/configuring.html#configuring-rules).

Scsslint also has [severities](https://github.com/brigade/scss-lint#severities) and it got [a mention](https://github.com/stylelint/stylelint/issues/1#issuecomment-67737941) over in the original Rules thread.

I don't know enough about this to decide whether the advantages of adding severities (e.g. easier integration into build systems) out weigh the disadvantages (e.g. more complexity), or not. Does anyone have a strong opinion on this?

I don't think we need to resolve this now though as this issue shouldn't affect how we write the rules themselves, right? Is it something that can be dealt with entirely in the [stylelint index.js](https://github.com/stylelint/stylelint/blob/master/src/index.js)? If so, shall we carry on as we were and just keep this issue hanging around for future reference?

**Comment by MoOx** [comment]
> If a power user is building on top a recommended styleguide then they don't need to worry about maintaining that part of the rules configuration.

+1 for this

> Does anyone have a strong opinion on this?

Not sure it will be that complex to handle with numbers.

> I don't think we need to resolve this now though as this issue shouldn't affect how we write the rules themselves, right? Is it something that can be dealt with entirely in the stylelint index.js?

If we need to choose warning vs errors we will need somewhere to adjust the behavior of a plugin. Maybe we can pass severity has an option if > 0 then save the severity in the postcss message. Then consumers that will read results (stylelint index.js?) will behave differently depending on severities.

That said, I am not sure how we make this works with simple plugin like postcss-log-warnings.
Initially I asked for [an api with log level built in](https://github.com/postcss/postcss/issues/177#issuecomment-73187756), but @ai [kind of refused this](https://github.com/postcss/postcss/issues/177#issuecomment-77554469) (I say "kind of" because object containing messages is called "messages" but public API is using the working "warnings" only, so not sure what we can do with that api).

**Comment by jeddy3** [comment]
> I am not sure how we make this works with simple plugin like postcss-log-warnings.

So, shall we hold off on this issue for now then, and continue to focus on getting the simplest/temporary version working (a postcss plugin that can be used with postcss-log-warnings to log warnings to console)?

I'll keep this issue open so anyone else can contribute to the discussion, or if they fancy having a go at coding a solution.

**Comment by MoOx** [comment]
> So, shall we hold off on this issue for now then, and continue to focus on getting the simplest/temporary version working

lgtm

**Comment by davidtheclark** [comment]
Am I oversimplifying in thinking that the difference between 1 and 2 is just the exit code if lint problems are reported (1 = exit code success, 0; 2 = exit code error, 1+)?

Just trying to understand.

I think it makes sense that stylelint couldn't force postcss-log-warnings to exit with an error or not an error. When the user configures logWarnings, they get to decide whether warnings matter. Stylelint would need its own reporter to force exit codes, I think.

**Comment by ai** [comment]
I think that error exit code should be generated by stylelint too.

I think we can have a API link in `gulp-eslint`:

```js
postcss()
    .use( stylelint() )
    .use( logWarnings() )
    .use( stylelint.failOnError() )
```

**Comment by davidtheclark** [comment]
Nice idea @ai

**Comment by MoOx** [comment]
We should not handle exit code, only runners should ! (ref http://eslint.org/docs/rules/no-process-exit )
At the end, I think stylelint tool will not be just a postcss plugin so we wont have to rely on postcss-log-warnings.
Also we have to think to future integrations with stuff like webpack that provide api to emit errors/warnings.

**Comment by MoOx** [comment]
ESLint has a bin runner & a node api. For my concern I use the webpack loader. So we can have multiples way to use the linter depending on the worflow, env etc. So we have to provide a way without just using exit() but messages with severity flags.

**Comment by davidtheclark** [comment]
So you're saying we'd have a severity flag on the message, then
- the stylelint CLI (bin runner) would know how to deal with them
- specialized stylelint plugins for webpack, gulp, whatever, would know how to deal with them
- non-specialized things that log PostCSS warning/error like gulp-postcss etc. or postcss-log-warnings would just print warnings but not interpret the severity flag.

Are we on the same page?

**Comment by MoOx** [comment]
Yes I think it's a solution.

> non-specialized things that log PostCSS warning/error like gulp-postcss etc. or postcss-log-warnings would just print warnings but not interpret the severity flag.

We could enhance those to make them better :)

**Comment by ai** [comment]
Why we can't just throw a error from `stylelint.failOnError`?

**Comment by MoOx** [comment]
Error is one severity level, we might have info, warnings or errors (recoverable error, so throwing is not useful).

**Comment by codemakerlive** [comment]
stdout would be an object with exit code and output string. Whatever works with PostCSS directly handles this object.

Fully passed lint is 0 (success), doesn't really need an output string - perhaps just "SUCCESS"

Warning exit code is 0 (success) with information for all warnings encountered provided in the output string.

Error exit occurs on first error encountered, is error code 1 (general error), with information on the error encountered in the exit string.

Filesystem errors and other exceptional errors could throw (some tools do this) or could return non-0 error, perhaps as per the following:
http://gsp.com/cgi-bin/man.cgi?section=3&topic=sysexits
http://www.tldp.org/LDP/abs/html/exitcodes.html
http://www.linuxjournal.com/article/10844

Only errors return non-0. That way when output from stylelint is piped to another tool it does not break/halt that piping.

Not sure why you need an info level? Just warnings or errors.

**Comment by stela5** [comment]
I agree with @codemakerlive -- only errors return non-0

It would be nice to have a .stylelintrc-ish config for *quiet* mode that only reports errors, not warnings

**Comment by davidtheclark** [comment]
Good idea @stela5. I made an issue for it: #383 .

**Comment by jeddy3** [comment]
@codemakerlive Thanks for your input btw. I believe a lot of your suggestions have been incorporated into https://github.com/stylelint/stylelint/pull/380

**Comment by davidtheclark** [comment]
Done in https://github.com/stylelint/stylelint/pull/380
